text/0000-modular-crypto-service.md

```diff
@@ -0,0 +1,349 @@
+
+Command Pattern:
+- Ref to above possbile ways of implemention, in either of solution, there need any paramter to decide with crypto library should be used.
+
+Take publicKey To EncryptedPEM as sample:
+
+#### Here is logic need to be added as a global wrapper to load plugin
+currently, some part of logic is implemented here. https://github.com/Hyperledger-TWGC/fabric/tree/bccsp-gm
+```go
+	// to do package gm
+	swbccsp.AddWrapper(reflect.TypeOf(&sm2.SM2PublicKey{}), &sm2.SM2PublicKeyKeyVerifier{})
+	......
+	func InitCryptoProviders() {
+		lock.Lock()
+		defer lock.Unlock()
+		// init
+		// publicKeyToEncryptedPEM
+		puk2epem = make(map[reflect.Type]func(k interface{}, pwd []byte) ([]byte, error))
+
+		//default AddWrapper for ecdsa
+		puk2epem[reflect.TypeOf(&ecdsa.PublicKey{})] = ECDSApublicKeyToEncryptedPEM
+
+		//AddWrapper for sm2
+		puk2epem[reflect.TypeOf(&ccssm2.PublicKey{})] = sm2.SM2publicKeyToEncryptedPEM
+	}
+```
+Full changes here: https://github.com/SamYuan1990/fabric/blob/packageRefactor/bccsp/sw/new.go#L107-L217
+
+#### Here is the logic changed in current sw to fill with wrapper above. 
+```go
+package sw
+
+publicKeyToEncryptedPEM(k interface{}, pwd []byte){
+	for i,v := range GetfromWrapperMap() {
+		if k.type == i.type {
+			return do(k,v)
+		}
+	}
+	return "not supported"
+}
+```
+Full changes here: https://github.com/SamYuan1990/fabric/blob/packageRefactor/bccsp/sw/keys.go#L315-L322
+
+#### A sample function as specific crypto wrapper/provider
+```go
+package specific
+import specific_crypto_lib
+
+//example function
+publicKeyToEncryptedPEM(k interface{}, pwd []byte){
+    specific impl
+}
+```
+Full changes here: https://github.com/Hyperledger-TWGC/fabric/tree/bccsp-gm/bccsp/sm2
+
+Option One:
+With help for go plugin, we are able to do:
+- Impl an interface as struct in plugin code.
+- Load the type of the interface by reflect.
+- Invoking interfaces method which provided by plugin.
+- From raw type to create plugin struct.
+For details, please refer https://github.com/SamYuan1990/go-plugindemo/blob/poc/main.go#L84-L110
+
+```go
+AddWapperMap(path string){
+p, err := plugin.Open(path.so)
+	if err != nil {
+
+	panic(err)
+
+	}
+	
+	initFunc, err := p.Lookup("init")
+  obj := initFunc().()
+
+	f, err := p.Lookup("publicKeyToEncryptedPEM")
+	...
+
+	my_type=reflect.Typeof(&obj)
+	//default for all type replacement
+	GlobalCryptoWrapper[my_type] = f
+}
+```
+
+Then we can make different kind of plugin.so file with interface defined in BCCSP and make the crypto implmentation plugable by switching the file.
+
+Option Two:
+We don't use the go plugin, as the limitation for ex: https://github.com/golang/go/issues/31354
+We put the specific crypto package as a part of bccsp package and rebuild fabric with the specific crypto package.(as build tag)
+Full changes here: https://github.com/Hyperledger-TWGC/fabric/tree/bccsp-gm
+
+Option Three:
```

Comment:
@denyeart , thanks you pointed out this modular crypto may need take a look at msp interface/package on contributor meeting last year when we do the review.
pre discussed with TWGC member, I had some attempt with bccsp interface and msp interface refactor/decouple idea as 3rd option here.
Please take a look at and have arrange/schedule review on contributor meeting.